apps/settings/js/panels/firefox_sync/firefox_sync_panel.js

```diff
@@ -0,0 +1,282 @@
+     *     all the options back and refresh the last sync time.
+     *     We disable this button if no collection is selected.
+     *     + Last sync time. Show the last time a successful synchronization
+     *     was done. If no previous synchronization exists for that login,
+     *     we don't show this label.
+     *     + Collections switches. Allow the user to select which collections
+     *     she wants to keep synchronized. If no collection is selected, the
+     *     synchronization will be disabled.
+     *     + ToS and Privacy links.
+     */
+    onsyncchange(message) {
+      if (!message || !message.state) {
+        throw new Error('Missing sync state');
+      }
+      switch (message.state) {
+        case 'disabled':
+          this.showScreen(LOGGED_OUT_SCREEN);
+          this.clean();
+          break;
+        case 'enabled':
+          this.showScreen(LOGGED_IN_SCREEN);
+          this.showSyncNow();
+          this.showUser(message.user);
+          this.showLastSync(message.lastSync);
+          break;
+        case 'syncing':
+          this.showScreen(LOGGED_IN_SCREEN);
+          this.showSyncing();
+          break;
+        case 'errored':
+          LazyLoader.load('shared/js/sync/errors.js', () => {
+            if (message.error == ERROR_UNVERIFIED_ACCOUNT) {
+              this.showScreen(LOGGED_IN_SCREEN);
+              this.showUnverified(message.user);
+              return;
+            }
+            this.showScreen(LOGGED_OUT_SCREEN);
+            this.clean();
+          });
+          break;
+      }
+    },
+
+    addListener(element, event, listener) {
+      if (!element || !event || !listener) {
+        return;
+      }
+      element.addEventListener(event, this[listener]);
+    },
+
+    removeListener(element, event, listener) {
+      if (!element || !event || !listener) {
+        return;
+      }
+      element.removeEventListener(event, this[listener]);
+    },
+
+    loadElements(screen) {
+      ELEMENTS.forEach(element => {
+        var name = getElementName(element.selector);
+        if (element.screen == screen) {
+          this.elements[name] = this.panel.querySelector(element.selector);
+          this.addListener(this.elements[name],
+                           element.event, element.listener);
+          return;
+        }
+        this.removeListener(this.elements[name],
+                            element.event, element.listener);
+        this.elements[name] = null;
+      });
+    },
+
+    showScreen(screen) {
+      if (this.currentScreen == screen) {
+        return;
+      }
+      this.currentScreen = screen;
+      this.screens.loggedIn.hidden = (screen != LOGGED_IN_SCREEN);
+      this.screens.loggedOut.hidden = (screen != LOGGED_OUT_SCREEN);
+      this.loadElements(screen);
+    },
+
+    showSyncing() {
+      this.elements.syncNow.dataset.l10nId = 'fxsync-syncing';
+      ['syncNow',
```

Comment:
Take this traversal with setting disabled to true or false into a separate method.